`Primitives/MergedStream.cs(22,8): error CS8026: Feature 'readonly automatically implemented properties' is not available in C# 5. Please use language version 6 or greater. [/home/taol/source/cs/OpenRA/OpenRA.Game/OpenRA.Game.csproj]`

--- a/OpenRA.Game/TraitDictionary.cs
+++ b/OpenRA.Game/TraitDictionary.cs
@@ -11,8 +11,10 @@
 
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
 using System.Linq;
 using OpenRA.Primitives;
+using OpenRA.Support;
 
 namespace OpenRA
 {
@@ -120,6 +122,16 @@ namespace OpenRA
 		{
 			foreach (var t in traits)
 				t.Value.RemoveActor(a.ActorID);
+		}
+
+		public void ApplyToActorsWithTrait<T>(Action<Actor, T> action)
+		{
+			InnerGet<T>().ApplyToAll(action);
+		}
+
+		public void ApplyToActorsWithTraitTimed<T>(Action<Actor, T> action, string text)
+		{
+			InnerGet<T>().ApplyToAllTimed(action, text);
 		}
 
 		interface ITraitContainer
@@ -276,6 +288,32 @@ namespace OpenRA
 				var count = endIndex - startIndex;
 				actors.RemoveRange(startIndex, count);
 				traits.RemoveRange(startIndex, count);
+			}
+
+			public void ApplyToAll(Action<Actor, T> action)
+			{
+				for (var i = 0; i < actors.Count; i++)
+					action(actors[i], traits[i]);
+			}
+
+			public void ApplyToAllTimed(Action<Actor, T> action, string text)
+			{
+				var longTickThresholdInStopwatchTicks = PerfTimer.LongTickThresholdInStopwatchTicks;
+				var start = Stopwatch.GetTimestamp();
+				for (var i = 0; i < actors.Count; i++)
+				{
+					var actor = actors[i];
+					var trait = traits[i];
+					action(actor, trait);
+					var current = Stopwatch.GetTimestamp();
+					if (current - start > longTickThresholdInStopwatchTicks)
+					{
+						PerfTimer.LogLongTick(start, current, text, trait);
+						start = Stopwatch.GetTimestamp();
+					}
+					else
+						start = current;
+				}
 			}
 		}
 	}

--- a/OpenRA.Game/World.cs
+++ b/OpenRA.Game/World.cs
@@ -465,7 +465,7 @@ namespace OpenRA
 					foreach (var a in actors.Values)
 						a.Tick();
 
-				ActorsWithTrait<ITick>().DoTimed(x => x.Trait.Tick(x.Actor), "Trait");
+				ApplyToActorsWithTraitTimed<ITick>((Actor actor, ITick trait) => trait.Tick(actor), "Trait");
 
 				effects.DoTimed(e => e.Tick(this), "Effect");
 			}
@@ -477,7 +477,7 @@ namespace OpenRA
 		// For things that want to update their render state once per tick, ignoring pause state
 		public void TickRender(WorldRenderer wr)
 		{
-			ActorsWithTrait<ITickRender>().DoTimed(x => x.Trait.TickRender(wr, x.Actor), "Render");
+			ApplyToActorsWithTraitTimed<ITickRender>((Actor actor, ITickRender trait) => trait.TickRender(wr, actor), "Render");
 			ScreenMap.TickRender();
 		}
 
@@ -534,6 +534,11 @@ namespace OpenRA
 		public IEnumerable<TraitPair<T>> ActorsWithTrait<T>()
 		{
 			return TraitDict.ActorsWithTrait<T>();
+		}
+
+		public void ApplyToActorsWithTraitTimed<T>(Action<Actor, T> action, string text)
+		{
+			TraitDict.ApplyToActorsWithTraitTimed<T>(action, text);
 		}
 
 		public IEnumerable<Actor> ActorsHavingTrait<T>()

